### ISSUE TITLE
[Web] CanvasKit not rendering text

### ISSUE THREAD

**Issue by videah** [flutter/flutter#47041]
Text and Icon's aren't rendering in CanvasKit for me out of the box. Is there something I must
do first or is this a bug? Even adding a FontManifest doesn't seem to fix it.

Compiled without `--dart-define=FLUTTER_WEB_USE_SKIA=true`
![image](https://user-images.githubusercontent.com/4761786/70848860-28a00400-1e6f-11ea-8cfa-d39e0675d67f.png)

Compiled with `--dart-define=FLUTTER_WEB_USE_SKIA=true`
![image](https://user-images.githubusercontent.com/4761786/70848850-12924380-1e6f-11ea-93a4-22c87dee8b0a.png)

**Comment by iapicca** [comment]
Hi @videah 
can you please provide your updated  `flutter doctor -v`?
Thank you

**Comment by videah** [comment]
Hi @iapicca, apologies for not including it initially.
```
PS D:\Users\Ruairidh> flutter doctor -v
[√] Flutter (Channel beta, v1.12.13+hotfix.6, on Microsoft Windows [Version 10.0.18362.535], locale en-GB)
    • Flutter version 1.12.13+hotfix.6 at D:\Users\Ruairidh\flutter
    • Framework revision 18cd7a3601 (5 days ago), 2019-12-11 06:35:39 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at D:\Users\Ruairidh\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: D:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at D:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 3.4)
    • Android Studio at D:\Program Files\Android\Android Studio
    • Flutter plugin version 39.0.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[√] IntelliJ IDEA Community Edition (version 2019.2)
    • IntelliJ at D:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.3.2
    • Flutter plugin version 42.1.2
    • Dart plugin version 192.7761

[√] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 79.0.3945.79
    • Web Server • web-server • web-javascript • Flutter Tools

• No issues found!
```

**Comment by yjbanov** [comment]
Are you using custom fonts? We have an issue rendering custom fonts. If it's a custom font then this is a duplicate of https://github.com/flutter/flutter/issues/45507.

**Comment by videah** [comment]
No, this is out of the box with a new project. It does this with everything I try to build it with.

**Comment by yjbanov** [comment]
Thanks. We'll take a look.

**Comment by PonyCui** [comment]
No custom font to reproduce this issue. It's really difficult to render text on WebGL, cause renderer must load font before render. And we don't have a good way to fetch font from browser, so we have to load from a remote server such as GoogleFonts. Load font from remote is okay for latin languages, but the Chinese or Japanese font file is so big.
We need to find out another way to render text on WebGL. The most common way is render in 2D offscreen canvas, and then use it as a texture and render in 3D WebGL context. Most of WebGL game engine use this way.
And the text layout is also difficult to resolve, I think we should use the same way as HTML to measure and layout the text.

**Comment by yjbanov** [comment]
@PonyCui we were actually discussing rendering text into an offscreen 2d canvas recently. In addition to the benefits you described, this would also allow us to remove Harfbuzz and ICU, which collectively take about half of the CanvasKit WASM size (~1MB). However, before we can implement it we need to complete our canvas-based text layout. @mdebbar is working on this as we speak. Please stay tuned!

**Comment by PonyCui** [comment]
I have implemented. But not sure is it the right way to solve this issue, may you look at it?

## Repos

https://github.com/PonyCui/flutter/tree/web_skia_fontLoadingPatch

https://github.com/PonyCui/flutter_skia_font_demo

## Solution

There's two way to render text, use Html.Canvas drawText method, use Html.SVG and Html.Image and Html.Canvas drawImage method.

### Html.Canvas drawText

It's really easy to draw text with Canvas, and the Flutter Web has [implmented](https://github.com/flutter/engine/blob/5dc24e3a503610227dbd0e3f065f9ae69e712793/lib/web_ui/lib/src/engine/bitmap_canvas.dart#L472) it.

Also, the Flutter Web has [implmented](https://github.com/flutter/engine/blob/5dc24e3a503610227dbd0e3f065f9ae69e712793/lib/web_ui/lib/src/engine/text/paragraph.dart#L236) a text layout service using Html.

So, we just need to use Flutter Web `EngineParagraph` and `BitmapCanvas`, and send result to Skia, render it.

### Html.SVG and Html.Image and Html.Canvas drawImage

Why we need SVG? Because `Html.Canvas drawText` can only draw simple text.

Image with SVG source, is a hacking way to resolve this problem. [Code](https://github.com/PonyCui/flutter/blob/fed253253c9daa87d9bd04b0b9da3a68ba83204e/lib/web_ui/lib/src/engine/compositor/canvas.dart#L186)

We put all text html to SVG data url, set source to ImageElement, and draw ImageElement to a offscreen CanvasElement.

And then, send result to Skia, render it.

## Implement

The Flutter engine code has pushed to [https://github.com/PonyCui/flutter/tree/web_skia_fontLoadingPatch](https://github.com/PonyCui/flutter/tree/web_skia_fontLoadingPatch).

The Demo code is [here](https://github.com/PonyCui/flutter_skia_font_demo).

We also provide a pre-compile `flutter_web_sdk` and a pre-built `build/web` product in demo repo.

**Comment by yjbanov** [comment]
@PonyCui That's very impressive! You have the right idea. If we go for this approach we're likely to use 2d canvas instead of SVG. It is true that out of the box 2d canvas cannot render complex text. However, @mdebbar is working on algorithms that will allows us to use it for complex text, including multi-line, multi-style, and ellipsized paragraphs, as well as `WidgetSpan` (https://github.com/flutter/flutter/issues/42086). It will take some time to get there but it's likely a better long-term solution.

Having said that, we are considering using SVG for the HTML rendering backend in situations where we can avoid using 2d canvas. There the `<foreignObject>` trick you used should be handy.

**Comment by Bestfastfire** [comment]
(The text that still appears is an image) The same here, but it only happens when I try on cell phones, on the PC it's normal, it's funny because when the website is opened, everything appears normally, but when I focus on a text field and the keyboard opens, everything disappears, sometimes, if you type in a field with obscure setted, the text shows and disappears several times, it happens throughout the app at the same time, include it there, see that it is completely blank.

![WhatsApp Image 2020-05-05 at 22 15 16](https://user-images.githubusercontent.com/22732544/81230108-49582a00-8fc7-11ea-8f55-cc3f341b0c95.jpeg)

**Comment by Bestfastfire** [comment]
Updating, I found out that the bug is probably only on Android, I ran on any browsers on the iPhone and the problem doesn’t happen, both Chrome and Safari on Android occur the problem and even more interesting, what may be causing this is the text field  with obscure on, see:

![WhatsApp Video 2020-05-07 at 11 40 43](https://user-images.githubusercontent.com/22732544/81308065-c03d0380-9057-11ea-8e25-0e9bbfc28cc7.gif)

**Comment by Bestfastfire** [comment]
flutter doctor:

Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, v1.17.0, on Microsoft Windows [versÃ£o 10.0.18362.778], locale pt-BR)
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    X Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit visit
      https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions.
[√] Chrome - develop for the web
[√] Android Studio (version 3.4)
[√] Android Studio (version 3.6)
[√] VS Code (version 1.44.2)
[√] Connected device (2 available)

! Doctor found issues in 1 category.

**Comment by AlraziBashir** [comment]
I have recently attempted to compile my flutter web app using Compiled with --dart-define=FLUTTER_WEB_USE_SKIA=true similar to @videah (CanvasKit rendering) without any success. How would I switch back to compiling my Flutter Web app with DomCanvas rendering, since CanvasKit is useless at this point?

![Screen Shot 2020-06-11 at 12 09 47 PM](https://user-images.githubusercontent.com/53784936/84411896-727a6480-abdc-11ea-8267-f08f6e5889b0.png)


![Screen Shot 2020-06-11 at 12 10 33 PM](https://user-images.githubusercontent.com/53784936/84411987-9178f680-abdc-11ea-8db3-577480809639.png)

![Screen Shot 2020-06-11 at 12 12 16 PM](https://user-images.githubusercontent.com/53784936/84412136-cbe29380-abdc-11ea-98b5-c4eaca756c3a.png)

**Comment by ArchangelXu** [comment]
@yjbanov Is there any updates on this issue?

**Comment by iapicca** [comment]
Hi @videah 
testing the counter app with the latest master

<details><summary>screenshot</summary>

![image](https://user-images.githubusercontent.com/45637986/88765138-b8848b00-d17e-11ea-94d1-854c2b1cedee.png)


</details>


<details><summary>logs</summary>

```bash
 nevercode@Nevercodes-Air  ~/Desktop/projects/master  flutter_master run -d chrome --dart-define=FLUTTER_WEB_USE_SKIA=true                               1 ↵  2923  09:32:27
Downloading Web SDK...                                              6.4s
Launching lib/main.dart on Chrome in debug mode...
Running "flutter pub get" in flutter_tools...                       1.3s
Syncing files to device Chrome...
20,979ms (!)
Debug service listening on ws://127.0.0.1:60078/mTIJhchhi-0=

Warning: Flutter's support for web development is not stable yet and hasn't
been thoroughly tested in production environments.
For more information see https://flutter.dev/web

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".
Cannot find native JavaScript type (CanvasKit) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Error: Unsupported operation: Picture.toImage not yet implemented for CanvasKit and HTML
    at Object.throw_ [as throw] (http://localhost:60039/dart_sdk.js:4324:11)
    at _engine.CkPicture.new.toImage (http://localhost:60039/dart_sdk.js:140618:17)
    at DefaultShaderWarmUp.execute (http://localhost:60039/packages/flutter/src/painting/shader_warm_up.dart.lib.js:106:23)
    at execute.next (<anonymous>)
    at http://localhost:60039/dart_sdk.js:37349:33
    at _RootZone.runUnary (http://localhost:60039/dart_sdk.js:37203:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:60039/dart_sdk.js:32187:29)
    at handleValueCallback (http://localhost:60039/dart_sdk.js:32734:49)
    at Function._propagateToListeners (http://localhost:60039/dart_sdk.js:32772:17)
    at _Future.new.[_completeWithValue] (http://localhost:60039/dart_sdk.js:32615:23)
    at async._AsyncCallbackEntry.new.callback (http://localhost:60039/dart_sdk.js:32637:35)
    at Object._microtaskLoop (http://localhost:60039/dart_sdk.js:37464:13)
    at _startMicrotaskLoop (http://localhost:60039/dart_sdk.js:37470:13)
    at http://localhost:60039/dart_sdk.js:32989:9Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkParagraph) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check
Cannot find native JavaScript type (SkMaskFilter) for type check

Application finished.
```

</details>


<details><summary>doctor</summary>

```console
[✓] Flutter (Channel master, 1.21.0-6.0.pre.89, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.21.0-6.0.pre.89 at /Users/nevercode/development/flutter_master
    • Framework revision ddb6d9d913 (4 hours ago), 2020-07-28 19:56:03 -0700
    • Engine revision 2d8a00e21d
    • Dart version 2.10.0 (build 2.10.0-1.0.dev 24c7666def)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (3 available)
    • macOS (desktop)  • macos      • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```

</details>

isn't the cleanest run I've ever seen and for some reason the plus icon is a bus (?)
nevertheless both `text` and `icon` are rendered running canvas kit

the issue seems to be solved, I feel safe to close this issue,
if you disagree please write in the comments and I will reopen it. 
Thank you

@Bestfastfire your issue seems to be browser specific, 
you may want to open a dedicated issue for that
providing a minimal complete reproducible code sample.
